rgw: add 'bucket sync checkpoint' command to radosgw-admin

cbodley opened the pull request:
Adds a consistency checkpoint command for bucket sync. This fetches the state of the bucket's sources from a remote zone, then polls the local zone's sync status until it catches up.

This command is then used directly in our multisite tests to implement `zone_bucket_checkpoint()`.

```
[zone a1] $ radosgw-admin bucket sync checkpoint --bucket testbucket --source-zone a2
2020-08-05T13:44:00.975-0400 7fcb766b5980  1 waiting for incremental sync to catch up:
      local status: [00000000004.8.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:05.977-0400 7fcb766b5980  1 waiting for incremental sync to catch up:
      local status: [00000000004.8.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:10.980-0400 7fcb766b5980  1 waiting for incremental sync to catch up:
      local status: [00000000004.8.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:15.982-0400 7fcb766b5980  1 bucket sync caught up with source:
      local status: [00000000008.18.5]
    remote markers: [00000000008.18.5]
2020-08-05T13:44:15.982-0400 7fcb766b5980  0 bucket checkpoint complete
```

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

yehudasa commented on src/rgw/rgw_admin.cc:
```diff
@@ -2639,6 +2642,234 @@ static int bucket_sync_status(rgw::sal::RGWRadosStore *store, const RGWBucketInf
+  for (size_t i = 0; i < lhs.size(); ++i) {
+    const auto& l = BucketIndexShardsManager::get_shard_marker(lhs[i].inc_marker.position);
+    const auto& r = rhs.get(i, "");
+    if (l >= r) {
```
@cbodley I assume we have a few shards that are greater or equal and a others that are smaller, so maybe the logic needs to be reversed? That is if `(l < r)` return `true`.

yehudasa commented on src/rgw/rgw_admin.cc:
```diff
@@ -2639,6 +2642,234 @@ static int bucket_sync_status(rgw::sal::RGWRadosStore *store, const RGWBucketInf
+
+      pipe.source.bucket = source_bucket;
+      pipe.dest.bucket = info.bucket;
+      r = bucket_source_sync_checkpoint(store, zone, z->second, c->second,
```
@cbodley I'd split the checkpoint into two: first get all the remote markers of all sources, and only then wait. This way we reduce the wait on stuff that arrived when we were waiting on another source.

cbodley commented:
updates:
* moved checkpoint code into new `rgw_sync_checkpoint.*`
* fixed marker comparison logic
* enabled optional_yield for `remote_bilog_markers()`
* initial steps (fetching remote markers and reading bucket info for each source) are done in parallel using optional_yield

adamemerson commented:
Jenkins, recheck this please

adamemerson commented:
This needs a fix to pass run-cli-tests
```
diff --git a/src/test/cli/radosgw-admin/help.t b/src/test/cli/radosgw-admin/help.t
index 220d5da7b0c..3a868d7d4ba 100644
--- a/src/test/cli/radosgw-admin/help.t
+++ b/src/test/cli/radosgw-admin/help.t
@@ -29,6 +29,7 @@
     bucket chown               link bucket to specified user and update its object ACLs
     bucket reshard             reshard bucket
     bucket rewrite             rewrite all objects in the specified bucket
+    bucket sync checkpoint     poll a bucket's sync status until it catches up to its remote
     bucket sync disable        disable bucket sync
     bucket sync enable         enable bucket sync
     bucket radoslist           list rados objects backing bucket's objects
```

adamemerson commented:
It looks like Bucket Creation Time is still failing on Teuthology.

http://qa-proxy.ceph.com/teuthology/aemerson-2020-08-28_05:57:46-rgw-wip-aemerson-qa-distro-basic-smithi/5381836/teuthology.log

cbodley commented:
thanks @adamemerson! i'll push an update for `help.t`. the `test_bucket_creation_time` fix in https://github.com/ceph/ceph/pull/36743 hasn't merged yet

cbodley commented:
rebased and amended help.t accordingly

cbodley commented:
https://pulpito.ceph.com/cbodley-2020-09-03_19:12:11-rgw:multisite-wip-cbodley-testing-distro-basic-smithi/

cbodley commented:
this is ready for another round of review!

cbodley commented:
jenkins test api
